app/src/main/java/org/thoughtcrime/securesms/jobs/FastJobStorage.java

```diff
@@ -0,0 +1,396 @@
+    deleteJobs(Collections.singletonList(jobId));
+  }
+
+  @Override
+  public synchronized void deleteJobs(@NonNull List<String> jobIds) {
+    List<String> durableIds = new ArrayList<>(jobIds.size());
+    for (String id : jobIds) {
+      JobSpec job = getJobById(id);
+      if (job == null || !job.isMemoryOnly()) {
+        durableIds.add(id);
+      }
+    }
+
+    if (durableIds.size() > 0) {
+      jobDatabase.deleteJobs(durableIds);
+    }
+
+    Set<String> deleteIds = new HashSet<>(jobIds);
+
+    Iterator<JobSpec> jobIter = jobs.iterator();
+    while (jobIter.hasNext()) {
+      if (deleteIds.contains(jobIter.next().getId())) {
+        jobIter.remove();
+      }
+    }
+
+    for (String jobId : jobIds) {
+      constraintsByJobId.remove(jobId);
+      dependenciesByJobId.remove(jobId);
+
+      for (Map.Entry<String, List<DependencySpec>> entry : dependenciesByJobId.entrySet()) {
+        Iterator<DependencySpec> depedencyIter = entry.getValue().iterator();
+
+        while (depedencyIter.hasNext()) {
+          if (depedencyIter.next().getDependsOnJobId().equals(jobId)) {
+            depedencyIter.remove();
+          }
+        }
+      }
+    }
+  }
+
+  @Override
+  public synchronized @NonNull List<ConstraintSpec> getConstraintSpecs(@NonNull String jobId) {
+    return Util.getOrDefault(constraintsByJobId, jobId, new LinkedList<>());
+  }
+
+  @Override
+  public synchronized @NonNull List<ConstraintSpec> getAllConstraintSpecs() {
+    return Stream.of(constraintsByJobId)
+                 .map(Map.Entry::getValue)
+                 .flatMap(Stream::of)
+                 .toList();
+  }
+
+  @Override
+  public synchronized @NonNull List<DependencySpec> getDependencySpecsThatDependOnJob(@NonNull String jobSpecId) {
+    List<DependencySpec> layer = getSingleLayerOfDependencySpecsThatDependOnJob(jobSpecId);
+    List<DependencySpec> all   = new ArrayList<>(layer);
+
+    Set<String> activeJobIds;
+
+    do {
+      activeJobIds = Stream.of(layer).map(DependencySpec::getJobId).collect(Collectors.toSet());
+      layer.clear();
+
+      for (String activeJobId : activeJobIds) {
+        layer.addAll(getSingleLayerOfDependencySpecsThatDependOnJob(activeJobId));
+      }
+
+      all.addAll(layer);
+    } while (!layer.isEmpty());
+
+    return all;
+  }
+
+  private @NonNull List<DependencySpec> getSingleLayerOfDependencySpecsThatDependOnJob(@NonNull String jobSpecId) {
+    return Stream.of(dependenciesByJobId.entrySet())
+                 .map(Map.Entry::getValue)
+                 .flatMap(Stream::of)
+                 .filter(j -> j.getDependsOnJobId().equals(jobSpecId))
+                 .toList();
+  }
+
+  @Override
+  public @NonNull List<DependencySpec> getAllDependencySpecs() {
+    return Stream.of(dependenciesByJobId)
+                 .map(Map.Entry::getValue)
+                 .flatMap(Stream::of)
+                 .toList();
+  }
+
+  private JobSpec getJobById(@NonNull String id) {
+    for (JobSpec job : jobs) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **388**. At this line the following field is not protected by synchronization: `jobs`. Your code typically protects this field with synchronization on object: `FastJobStorage.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `FastJobStorage.this`. For example, like at lines: **86**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **55**, **78**, **86**, **191**, **220**, **246**, and **314**.